### ISSUE TITLE
General Hashing Function for Proto Object

### ISSUE THREAD

**Issue by terenc3t** [prysmaticlabs/prysm#1192]
We have seen the following [patterns](https://github.com/prysmaticlabs/prysm/blob/master/beacon-chain/core/validators/validator.go#L475-L479) all over the repo with our proto objects. It'd be nice to have a generalized hashing function which can take in any object that fulfills the proto interface and use that to hash the object. This function should reside in [hashutils](https://github.com/prysmaticlabs/prysm/tree/master/shared/hashutil) and should abstract the marshaling and then hashing similarly to the following:
```go

func HashProto(object proto.Message) ([32]byte,error) {
 marshalledObj := proto.Marshal(object)
 return Hash(marshalledObj)
}
```

**Comment by aunyks** [comment]
Would any existing structures need to be changed after this functionality is added?

**Comment by nisdas** [comment]
hey @aunyks , yes the idea is that we would not need separate hashing functions for different proto objects. So for example over here: 
https://github.com/prysmaticlabs/prysm/blob/master/beacon-chain/core/blocks/block.go#L18
https://github.com/prysmaticlabs/prysm/blob/master/beacon-chain/core/state/state.go#L142

we wouldn't need the two above functions, so a generalized hashing function would look like:
this:

`func HashProto(object proto.Message) ([32]byte,error) `

A PR would replace all usages of the specific hash functions with the generalized functions.

I have changed the above code example to be more accurate.

**Comment by aunyks** [comment]
Ah, I see. I'd like to tackle this!

**Comment by terenc3t** [comment]
Thanks Gerald! Assigned to you

**Comment by prestonvanloon** [comment]
This could also be better as a proto extension. It would be nice to just call `.Hash()` on our proto message.

**Comment by rauljordan** [comment]
Closed by #1414
